### PR TITLE
[SYCL][OpenCL] Port OpenCL plugin to Unified Runtime

### PR DESCRIPTION
**Pull request by fabiomestre**

### CODE REVIEW

**Review by jandres742 — APPROVED** [review]

File: sycl/plugins/unified_runtime/pi2ur.hpp [hunk 1]
```diff
@@ -3892,6 +3886,17 @@ inline pi_result piEventGetInfo(pi_event Event, pi_event_info ParamName,
   HANDLE_ERRORS(urEventGetInfo(UrEvent, PropName, ParamValueSize, ParamValue,
                                ParamValueSizeRet));
 
+  if (ParamName == PI_EVENT_INFO_COMMAND_EXECUTION_STATUS) {
```
[the remark this answers is not in the document]
Comment (fabiomestre) [reply sycl/plugins/unified_runtime/pi2ur.hpp]: That makes sense. I created a new PR for this change: https://github.com/intel/llvm/pull/10756

File: sycl/plugins/unified_runtime/ur/adapters/opencl/enqueue.cpp [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+
+UR_APIEXPORT ur_result_t UR_APICALL urEnqueueMemBufferWriteRect(
+    ur_queue_handle_t hQueue, ur_mem_handle_t hBuffer, bool blockingWrite,
+    ur_rect_offset_t bufferOrigin, ur_rect_offset_t hostOrigin,
+    ur_rect_region_t region, size_t bufferRowPitch, size_t bufferSlicePitch,
+    size_t hostRowPitch, size_t hostSlicePitch, void *pSrc,
+    uint32_t numEventsInWaitList, const ur_event_handle_t *phEventWaitList,
+    ur_event_handle_t *phEvent) {
+
+  UR_ASSERT(hQueue, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
```
[the remark this answers is not in the document]
Comment (fabiomestre) [reply sycl/plugins/unified_runtime/ur/adapters/opencl/enqueue.cpp]: Done :+1:

File: sycl/plugins/unified_runtime/ur/adapters/opencl/queue.cpp [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+  cl_platform_id CurPlatform;
+  CL_RETURN_ON_FAILURE_AND_SET_NULL(
+      clGetDeviceInfo(cl_adapter::cast<cl_device_id>(hDevice),
+                      CL_DEVICE_PLATFORM, sizeof(cl_platform_id), &CurPlatform,
+                      nullptr),
+      phQueue);
+
+  cl_command_queue_properties CLProperties =
+      convertURQueuePropertiesToCL(pProperties);
```
Comment (aarongreig) [inline sycl/plugins/unified_runtime/ur/adapters/opencl/queue.cpp:64]:
```suggestion
-      convertURQueuePropertiesToCL(pProperties);
+      pProperties ? convertURQueuePropertiesToCL(pProperties) : 0;
```
pProperties is optional

**Comment by fabiomestre** [comment]
I have rebased this PR on top of sycl branch. Would appreaciate if someone from @intel/llvm-reviewers-runtime could review it

**Review by aelovikov-intel — APPROVED** [review]
SYCL RT (non-PI/non-UR) part (which is basically just the test change) LGTM.

**Review by isaacault — APPROVED** [review]
Bindless Images (`adapters/*/image.*`) LGTM

**Review by Seanst98 — APPROVED** [review]

**Comment by fabiomestre** [comment]
@intel/llvm-gatekeepers I think that this PR is ready to be merged.

**Comment by steffenlarsen** [comment]
We are having some issues with our Windows CI and for this particular PR it would be best to have it passing on Windows as well. I will retrigger it and hopefully it will work this time around.

**Comment by fabiomestre** [comment]
> We are having some issues with our Windows CI and for this particular PR it would be best to have it passing on Windows as well. I will retrigger it and hopefully it will work this time around.

It just came to my attention that this revert was merged yesterday: https://github.com/intel/llvm/pull/11227
The OpenCL port has dependencies on the commit that got reverted. So it cannot be merged until a solution is found for that issue.

**Comment by fabiomestre** [comment]
@intel/llvm-gatekeepers This PR should now again be ready to be merged.

**Pull request merged**